File: sdks/ios/Sources/XMTPiOS/Codecs/DeleteMessageCodec.swift
import Foundation

public let ContentTypeDeleteMessageRequest = ContentTypeID(
	authorityID: "xmtp.org",
	typeID: "deleteMessage",
	versionMajor: 1,
	versionMinor: 0,
)

/// Represents a request to delete a message.
/// This content type is used to request deletion of a specific message in a conversation.
/// The message will be deleted for all participants.
public struct DeleteMessageRequest: Codable, Equatable {
	/// The ID of the message to delete
	public var messageId: String

	public init(messageId: String) {
		self.messageId = messageId
	}
}

public struct DeleteMessageCodec: ContentCodec {
	public typealias T = DeleteMessageRequest

	public init() {}

	public var contentType: ContentTypeID = ContentTypeDeleteMessageRequest

	public func encode(content: DeleteMessageRequest) throws -> EncodedContent {
		let ffi = FfiDeleteMessage(
			messageId: content.messageId,
		)
		return try EncodedContent(serializedBytes: encodeDeleteMessage(request: ffi))
	}

	public func decode(content: EncodedContent) throws -> DeleteMessageRequest {
		let decoded = try decodeDeleteMessage(bytes: content.serializedBytes())
		return DeleteMessageRequest(
			messageId: decoded.messageId,
		)
	}

	public func fallback(content _: DeleteMessageRequest) throws -> String? {
		nil
	}

	public func shouldPush(content _: DeleteMessageRequest) throws -> Bool {
		false
	}
}

File: library/src/main/java/org/xmtp/android/library/codecs/DeleteMessageCodec.kt
package org.xmtp.android.library.codecs

/**
 * Represents a request to delete a message.
 *
 * This content type is used to request deletion of a specific message in a conversation.
 * The message will be deleted for all participants.
 *
 * @property messageId The ID of the message to delete
 */
data class DeleteMessageRequest(
    val messageId: String,
)

val ContentTypeDeleteMessageRequest =
    ContentTypeIdBuilder.builderFromAuthorityId(
        "xmtp.org",
        "deleteMessage",
        versionMajor = 1,
        versionMinor = 0,
    )

data class DeleteMessageCodec(
    override var contentType: ContentTypeId = ContentTypeDeleteMessageRequest,
) : ContentCodec<DeleteMessageRequest> {
    override fun encode(content: DeleteMessageRequest): EncodedContent {
        val ffi =
            uniffi.xmtpv3.FfiDeleteMessage(
                messageId = content.messageId,
            )

        return EncodedContent.parseFrom(
            uniffi.xmtpv3.encodeDeleteMessage(ffi),
        )
    }

    override fun decode(content: EncodedContent): DeleteMessageRequest {
        val decoded = uniffi.xmtpv3.decodeDeleteMessage(content.toByteArray())

        return DeleteMessageRequest(
            messageId = decoded.messageId,
        )
    }

    override fun fallback(content: DeleteMessageRequest): String? = null

    override fun shouldPush(content: DeleteMessageRequest): Boolean = false
}
